networks/3a_validate_array_vs_rnaseq_modules.py

```diff
@@ -0,0 +1,217 @@
+
+# Clustering using hierarchal clustering
+if cluster_method == "hierarchal":
+    pao1_array_clustering = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=hier_threshold, linkage=link_dist
+    ).fit(ppao1_array_corr_log_spell)
+    pao1_rnaseq_clustering = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=hier_threshold, linkage=link_dist
+    ).fit(pao1_rnaseq_corr_log_spell)
+
+# Clustering using affinity propogation
+if cluster_method == "affinity":
+    pao1_array_clustering = AffinityPropagation(random_state=0).fit(
+        pao1_array_corr_log_spell
+    )
+    pao1_rnaseq_clustering = AffinityPropagation(
+        random_state=1, damping=affinity_damping
+    ).fit(pao1_rnaseq_corr_log_spell)
+
+# +
+# Get module membership for a single threshold
+# Format and save output to have columns: gene_id | group_id
+pao1_array_membership_df = pd.DataFrame(
+    data={"module id": pao1_array_clustering.labels_},
+    index=pao1_array_corr_log_spell.index,
+)
+
+pao1_array_membership_df["module id"].value_counts()
+# -
+
+pao1_array_membership_df.head()
+
+# +
+pao1_rnaseq_membership_df = pd.DataFrame(
+    data={"module id": pao1_rnaseq_clustering.labels_},
+    index=pao1_rnaseq_corr_log_spell.index,
+)
+
+pao1_rnaseq_membership_df["module id"].value_counts()
+# -
+
+pao1_rnaseq_membership_df.head()
+
+# ## Compare composition of modules
+#
+# For a given array module, are the genes within 1 module in the RNA-seq compendium?
+
+# +
+array_module_mapping = {}
+for grp_id, grp in pao1_array_membership_df.groupby("module id"):
+    grp_mapped = pao1_rnaseq_membership_df.loc[grp.index]
+    array_module_mapping[grp_id] = list(grp_mapped["module id"].unique())
+
+array_module_mapping
+
+# +
+# Are there any array modules that map to a single rnaseq module?
+consistent_array_modules = []
+for array_module_id, list_rnaseq_module_ids in array_module_mapping.items():
+    if len(list_rnaseq_module_ids) == 1:
+        print(array_module_id, list_rnaseq_module_ids)
+        consistent_array_modules.append(array_module_id)
+
+print(len(consistent_array_modules) / len(array_module_mapping))
+
+# +
+rnaseq_module_mapping = {}
+for grp_id, grp in pao1_rnaseq_membership_df.groupby("module id"):
+    grp_mapped = pao1_array_membership_df.loc[grp.index]
+    rnaseq_module_mapping[grp_id] = list(grp_mapped["module id"].unique())
+
+rnaseq_module_mapping
+
+# +
+# Are there any rnaseq modules that map to a single array module?
+consistent_rnaseq_modules = []
+for rnaseq_module_id, list_array_module_ids in rnaseq_module_mapping.items():
+    if len(list_array_module_ids) == 1:
+        print(rnaseq_module_id, list_array_module_ids)
+        consistent_rnaseq_modules.append(rnaseq_module_id)
+
+print(len(consistent_rnaseq_modules) / len(rnaseq_module_mapping))
+# -
+
+# **Observation:**
+#
+# * Only ~3% of modules are consistent using RNA-seq and array data. We would have expected more of an overlap. What is causing this?
```

Comment:
The first thing I would do is to use exactly the same parameter set for both data types. I think you are using a different dumping factor for Affinity propagation. Not sure, though, if you already analyzed this and have a reason for doing it.

The second thing is that you are running the algorithm only once if I'm not mistaken. I'm not very familiar with affinity propagation, but the fact that it supports a `random_state` parameter makes me think that if you run it several times, it will produce different partitions.

Third, what's the result with the other clustering algorithms? Both of them, hierarchical clustering and DBSCAN, have the advantage of being deterministic, so it would be interesting to see how they performed. Other algorithms like Spectral clustering allow you to specify `n_init`, such as 100 or 200, meaning that the algorithm will be run 100 or 200 times and then it will return the "best" partition. That would be probably a better strategy than just taking a random run (such as with Affinity propagation).

A fourth point would be to report also the similarity given by measures like ARI.

Reply:
Thank you! I really appreciate all these very helpful feedback. I'll definitely update my clustering and validation steps accordingly moving forward.